### ISSUE TITLE
RFC: support type-level fixpoint

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#1028]
Follow-up to the discussion from https://github.com/dhall-lang/dhall-lang/issues/1027.
The idea is that we could add support for the type of fixpoints we do eg. in `Prelude.JSON.Type`. The idea would be to write
```dhall
List = ∀(a : Type) →
       fix(list_a : Type) →
       < cons : { head : a, tail : list_a } | nil >
```
instead of
```dhall
List = ∀(a : Type) →
       ∀(list_a : Type) →
       (< cons : { head : a, tail : list_a } | nil > → list_a) → list_a
```

It's not exactly clear how it would work, but I would like it if we could have normal-looking constructors and `merge` on those types, eg.:
```dhall
let Nat = fix(nat : Type) → < zero | succ: nat >
let nat_three = Nat.succ (Nat.succ (Nat.succ Nat.zero))
let actual_three = merge {
		zero = 0,
		succ = \(x: Natural) -> x+1  -- note that `x` is an actual number here
	} nat_three
let _ = assert : actual_three === 3
```

**Comment by Nadrieril** [comment]
Let's define a magic constructor `Fix` that works like Haskell's `data Fix f = f (Fix f)`. This would be the constructor for the type `fix(x: T) -> ...`. @Gabriel439 proposes a `fold` operator:
```dhall
fold : (f r → r) → (fix(x: T) -> f x) → r
```
In sequents:
```
Γ₀ ⊢ u : fix(x: T) -> e   Γ₀ ⊢ f : s → r   e[x ≔ r] = s
───────────────────────────────────────────────────────
Γ₀ ⊢ fold f u : r

u ⇥ Fix(v)   f (magic_map (fold f) v) ⇥ b
──────────────────────────────────
fold f u ⇥ b
```
Note that we would need a magic `map` operator that works like Haskell's `fmap`, because we need:
```dhall
magic_map (fold f) : e[x ≔ fix(x: T) -> e)] → e[x ≔ r]
```
If we give a name to the functor `F = \(x: T) -> e`, that would look like
```dhall
magic_map : (x -> y) -> F x → F y
```
We would have to restrict `F` to be a functor, which is probably the main difficulty in this proposal.


If we want nice constructors/merge, we also need to special-case unions:
```
e ⇥ fix(rec : Type) → u₀   u₀ ⇥ < xs… >  u₀[rec ≔ e] = u₁  Fix(u₁.x v) ⇥ b
──────────────────────────────────────────────────────────────────────────
e.x v ⇥ b

u ⇥ Fix(_)   fold (merge t) u ⇥ b
─────────────────────────────────
merge t u ⇥ b
```
That way `Nat.succ x` normalizes to `Fix(< zero | succ: Nat >.succ x)`, and `merge` unpacks the recursion invisibly.

**Comment by Gabriel439** [comment]
@Nadrieril: First, I think we should call this `Recursive` or something similar that is more descriptive than `Fix`, for two reasons:

* "Fix" doesn't distinguish between a least fixed point and a greatest fixed point, which would be encoded differently and have different semantic properties

  Their respective encodings in Dhall:

  ```dhall
  -- Least fixed point
  let Recursive
      : (Type → Type) → Type
      = λ(f : Type → Type) → ∀(r : Type) → (f r → r) → r

  -- Greatest fixed point
  let Corecursive
      : (Type → Type) → Type
      = λ(f : Type → Type) → ∀(r : Type) → (∀(s : Type) → s → (s → f s) → r) → r

  in  {=}
  ```

* I believe more people will understand the connotation of "Recursive" than the connotation of "Fix" (since fewer people know that Fix stands for fixed point and fewer people know that fixed points are a way of encoding recursion)

Also, what I was proposing was something like this:

```dhall
let List : Type → Type =
      ∀(a : Type) →
        Recursive (λ(list : Type) → < Cons : { head : a, tail : list } | Nil >)
```

In other words, `Recursive` would be a built-in (not a keyword), that has type `(Type → Type) → Type`.

Also, the type of the `fold` keyword would be:

```dhall
Γ ⊢ a :⇥ Recursive f   Γ ⊢ b :⇥ f r → r
────────────────────────────────────────────
Γ ⊢ fold a b : r
```

... although, thinking about it more, perhaps it would make more sense to have `Recursive/fold` and `Recursive/build` keywords (to leave the door open for `Corecursive/{fold,build}`).  Their implementation would be really simple, `Recursive/build` would be an inert constructor for a `Recursive` type, and `Recursive/fold` would remove the `Recursive` constructor:

```
Γ ⊢ a :⇥ forall (r : Type) → (f r → r) → r
────────────────────────────────────────────
Γ ⊢ Recursive/build a : Recursive f

Γ ⊢ b :⇥ Recursive f
────────────────────────────────────────────
Γ ⊢ Recursive/fold b : forall (r : Type) → (f r → r) → r


a₀ ⇥ a₁
────────────────────────────────────────────
Recursive/build a₀ ⇥ Recursive a₁

b ⇥ Recursive/build a
────────────────────────────────────────────
Recursive/fold b ⇥ a
```

I think that would be easier for people to implement because there is no magic involved: `Recursive` just plays the role of a "newtype" around `forall (r : Type) → (f r → r) → r` and `Recursive/build` and `Recursive/fold` are the newtype wrapper and unwrapper, respectively.

**Comment by Nadrieril** [comment]
Oh, that's definitely cleaner. I should have guessed we can avoid `map` given that we don't need it with the Church encoding. Glad we can avoid that.

Let me try again the syntactic sugar for recursive unions:
```
e ⇥ Recursive f
Γ ⊢ (f e).x v : _
────────────────────────────────────────────
Γ ⊢ e.x v : e

e ⇥ Recursive f
Recursive/build (\(r: Type) -> \(g: f r -> r) -> g ((f r).x (Recursive/fold v r g))) ⇥ a
────────────────────────────────────────────────────────────────────────────────────────
e.x v ⇥ a

Γ ⊢ u :⇥ Recursive f
Γ ⊢ (\(v: f T) -> merge t v : T) : _
────────────────────────────────────────────
Γ ⊢ (merge t u : T) : T

u ⇥ Recursive/build _   Recursive/fold u T (merge t) ⇥ a
────────────────────────────────────────────
merge t u : T ⇥ a
```
It looks like I need a type annotation on `merge`, otherwise I can't infer what type to apply `f` to... That could be worked around by inspecting `t` but ugh.

**Comment by singpolyma** [comment]
This proposal is, frankly, horrifying to me.  I mean, I *guess* it's less bad than #1027, but only a little.  A whole language feature, to implement a macro for an advanced use case that will usually live in a library and not in average end-user code?  No thanks.

**Comment by Nadrieril** [comment]
I do agree that it's a very niche usecase, even more so if we adopt #1027

**Comment by winitzki** [comment]
I don't see lots of advantages in introducing a built-in for something that's straightforwardly implemented via Church encoding. The only thing is that we will be able to create tooling to export these structures to JSON / YAML more easily.

```dhall
let Recursive
   : ∀(f : Type → Type) → Type
   = λ(f : Type → Type) → ∀(r : Type) → (f r → r) → r
```
After this, `Recursive/fold` and `Recursive/build` are easy.

The only disadvantage is the ugly Church encoding type `∀(r : Type) → (f r → r) → r` that we would have to look at, all the time. Instead we just want to write `Recursive f` and we don't want Dhall to write out the full Church encoding. That's the extent of this proposal.

It still won't be easy to deal with this type. There will need to be special forms for `fmap`, for constructors (`fix`), for pattern-matching (`unfix`), and for `hoist` (transformation from `∀(t : Type) → f t → g t` to `Recursive f → Recursive g`.

One possible advantage is that Dhall implementations could substitute more efficient code for `Recursive f`.
